akka-projection-cassandra/src/main/scala/akka/projection/cassandra/scaladsl/CassandraProjection.scala

```diff
@@ -55,23 +55,28 @@ object CassandraProjection {
   def atMostOnce[Offset, Envelope](
       projectionId: ProjectionId,
       sourceProvider: SourceProvider[Offset, Envelope],
-      handler: Handler[Envelope]): CassandraProjection[Envelope] =
-    new CassandraProjectionImpl(
-      projectionId,
-      sourceProvider,
-      CassandraProjectionImpl.AtMostOnce,
-      settingsOpt = None,
-      handler)
+      handler: Handler[Envelope]): AtMostOnceCassandraProjection[Envelope] =
+    new AtMostOnceCassandraProjectionImpl(projectionId, sourceProvider, projectionSettings = None, handler)
 }
 
 trait CassandraProjection[Envelope] extends Projection[Envelope] {
 
-  override def withSettings(settings: ProjectionSettings): CassandraProjection[Envelope]
-
   /**
    * For testing purposes the offset table can be created programmatically.
    * For production it's recommended to create the table with DDL statements
    * before the system is started.
    */
   def createOffsetTableIfNotExists()(implicit systemProvider: ClassicActorSystemProvider): Future[Done]
 }
+
+trait AtLeastOnceCassandraProjection[Envelope] extends CassandraProjection[Envelope] {
+  override def withSettings(settings: ProjectionSettings): AtLeastOnceCassandraProjection[Envelope]
+
+  def withAtLeastOnceSettings(settings: AtLeastOnceSettings): AtLeastOnceCassandraProjection[Envelope]
+  def withSaveOffsetAfterEnvelopes(afterEnvelopes: Int): AtLeastOnceCassandraProjection[Envelope]
+  def withSaveOffsetAfterDuration(afterDuration: FiniteDuration): AtLeastOnceCassandraProjection[Envelope]
+}
```

Comment:
This is on the right track, but I think it's more changes than necessary in the `CassandraProjectionImpl`.
I was thinking along the lines...

`def atLeastOnce` returns `AtLeastOnceCassandraProjection`:
```
  def atLeastOnce[Offset, Envelope](
      projectionId: ProjectionId,
      sourceProvider: SourceProvider[Offset, Envelope],
      handler: Handler[Envelope]): AtLeastOnceCassandraProjection[Envelope] =
    new CassandraProjectionImpl(
      projectionId,
      sourceProvider,
      CassandraProjectionImpl.AtLeastOnce(None, None),
      settingsOpt = None,
      handler)
```

Note that the `AtLeastOnce` is starting with None, None, indicating that no specific parameters have been chosen, and defaults should be used.

```
trait AtLeastOnceCassandraProjection[Envelope] extends CassandraProjection[Envelope] {
  def withSaveOffset(afterEnvelopes: Int, orAfterDuration: FiniteDuration): AtLeastOnceCassandraProjection[Envelope]

  override def withSettings(settings: ProjectionSettings): AtLeastOnceCassandraProjection[Envelope]
}
```

and the implementation:
```
@InternalApi private[akka] class CassandraProjectionImpl[Offset, Envelope](
    override val projectionId: ProjectionId,
    sourceProvider: SourceProvider[Offset, Envelope],
    strategy: CassandraProjectionImpl.Strategy,
    settingsOpt: Option[ProjectionSettings],
    handler: Handler[Envelope])
    extends javadsl.AtLeastOnceCassandraProjection[Envelope]
    with scaladsl.AtLeastOnceCassandraProjection[Envelope] {
  import CassandraProjectionImpl._
  import HandlerRecoveryImpl.applyUserRecovery

  override def withSaveOffset(
      afterEnvelopes: Int,
      orAfterDuration: FiniteDuration): CassandraProjectionImpl[Offset, Envelope] = {
    new CassandraProjectionImpl(
      projectionId,
      sourceProvider,
      AtLeastOnce(afterEnvelopes, orAfterDuration),
      settingsOpt,
      handler)
  }
```

Then when running the projection it will grab the parameters from the AtLeastOnce strategy, or fallback to ProjectionSettings if None.

```
trait ProjectionSettings {

  def atLeastOnceSaveOffsetAfterEnvelopes: Int
  def atLeastOnceSaveOffsetAfterDuration: FiniteDuration
```

The ProjectionSetting is only defining the default values, read from config. We should not have any with methods for these parameters in ProjectionSetting because it would be confusing to be able to define it in two ways or define it when not used (e.g. atMostOnce).

Reply:
Thanks for the direction (and detail).  Instead of reverting the strategy re-factory I created a new PR https://github.com/akka/akka-projection/pull/175

Configuration aside, WDYT about factoring out the strategy implementations?